clients/vscode/src/lsp/WorkspaceFeature.ts

```diff
@@ -0,0 +1,141 @@
+import EventEmitter from "events";
+import { ApplyWorkspaceEditParams, ApplyWorkspaceEditRequest } from "tabby-agent";
+import {
+  BaseLanguageClient,
+  FeatureState,
+  ShowDocumentParams,
+  ShowDocumentRequest,
+  ShowDocumentResult,
+  StaticFeature,
+  TextEdit,
+} from "vscode-languageclient";
+import { Disposable, Position, Range, TextDocument, TextEditorEdit, window, workspace } from "vscode";
+import { diffLines } from "diff";
+export class WorkSpaceFeature extends EventEmitter implements StaticFeature {
+  private registration: string | undefined = undefined;
+  private disposables: Disposable[] = [];
+
+  constructor(private readonly client: BaseLanguageClient) {
+    super();
+  }
+  getState(): FeatureState {
+    return { kind: "static" };
+  }
+
+  fillInitializeParams() {
+    // nothing
+  }
+
+  fillClientCapabilities(): void {
+    // nothing
+  }
+
+  preInitialize(): void {
+    // nothing
+  }
+
+  initialize(): void {
+    this.disposables.push(
+      this.client.onRequest(ApplyWorkspaceEditRequest.type, (params: ApplyWorkspaceEditParams) => {
+        return this.handleApplyWorkspaceEdit(params);
+      }),
+      this.client.onRequest(ShowDocumentRequest.type, (params: ShowDocumentParams) => {
+        return this.handleRevealEditorRange(params);
```

Comment:
I thought this standard LSP method should already be implemented by VSCode, so we do not need to implement it again. Could you help confirm this?

Reply:
ye sure, give me a second